core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -36,25 +36,41 @@ class Replica(val brokerId: Int,
   // for local replica it is the log's end offset, for remote replicas its value is only updated by follower fetch
   @volatile private[this] var logEndOffsetMetadata: LogOffsetMetadata = LogOffsetMetadata.UnknownOffsetMetadata
 
+  // The log end offset value at the time the leader receives last FetchRequest from this follower.
+  // This is used to determine the last catch-up time of the follower
+  @volatile var lastFetchLeaderLogEndOffset: Long = 0L
```

Comment:
I agree that we should by default consider a replica to be out-of-sync. I have made the following two fixes. I have kept `lastFetchLeaderLogEndOffset` as 0L. This is because the first fetch request from a follower will set `lastCaughtUpTimeMsUnderlying` to `lastFetchTimeMs`, which is initialized to 0L in Replica.java.

- Initialize `lastCaughtUpTimeMsUnderlying` to `AtomicLong(0L)` in Replica.java
- When broker receives LeaderAndIsrRequest to make it leader for a partition, reset `lastCaughtUpTimeMsUnderlying` of all replicas of this partition to 0L.